lib/si-layer-cache/src/hybrid_cache.rs

```diff
@@ -0,0 +1,174 @@
+                // todo: bad clone here
+                MaybeDeserialized::DeserializedValue(v) => Some(v.clone()),
+                MaybeDeserialized::RawBytes(bytes) => {
+                    // If we fail to deserialize the raw bytes for some reason, pretend that we never
+                    // had the key in the first place, and also remove it from the cache.
+                    match serialize::from_bytes_async::<V>(bytes).await {
+                        Ok(deserialized) => {
+                            self.insert(key.into(), deserialized.clone());
+                            Some(deserialized)
+                        }
+                        Err(e) => {
+                            error!(
+                        "Failed to deserialize stored bytes from memory cache for key ({:?}): {}",
+                        key,
+                        e
+                    );
+                            self.remove(key);
+                            None
+                        }
+                    }
+                }
+            },
+
+            _ => None,
+        }
+    }
+
+    pub fn insert(&self, key: Arc<str>, value: V) {
+        self.cache
+            .insert(key, MaybeDeserialized::DeserializedValue(value));
+    }
+
+    pub fn insert_raw_bytes(&self, key: Arc<str>, raw_bytes: Vec<u8>) {
+        self.cache
+            .insert(key, MaybeDeserialized::RawBytes(raw_bytes));
+    }
+
+    pub fn remove(&self, key: &str) {
+        self.cache.remove(key);
+    }
+
+    pub fn contains(&self, key: &str) -> bool {
+        self.cache.contains(key)
+    }
+
+    pub async fn close(&self) -> LayerDbResult<()> {
+        self.cache
+            .close()
+            .await
+            .map_err(|e| LayerDbError::Foyer(e.into()))?;
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct CacheConfig {
+    disk_capacity: usize,
+    disk_path: PathBuf,
+    disk_admission_rate_limit: usize,
+    disk_buffer_size: usize,
+    disk_buffer_flushers: usize,
+    memory: u64,
+    memory_percentage: f32,
+    name: String,
+}
+
+impl CacheConfig {
+    // set the total size of the disk cache for this instance of the cache
+    pub fn with_disk_capacity(mut self, capacity: usize) -> Self {
+        self.disk_capacity = capacity;
+        self
+    }
+
+    // set the percentage of total memory
+    pub fn with_memory_percentage(mut self, memory_percentage: f32) -> Self {
+        self.memory_percentage = memory_percentage;
+        self
+    }
+
+    // append an additional path to the existing disk path
+    pub fn with_path_join(mut self, path: impl AsRef<Path>) -> Self {
+        self.disk_path = self.disk_path.join(path);
+        self
+    }
+}
+
+impl Default for CacheConfig {
+    fn default() -> Self {
+        let sys = sysinfo::System::new_all();
+        let path = tempfile::TempDir::with_prefix_in("default-cache-", "/tmp")
+            .expect("unable to create tmp dir for layerdb")
```

Comment:
I *think* we're okay with this potential panic as we're constructing configs and even their caches early at boot time for our services (i.e. no delayed time bomb). Alternatively you may be able to do something like `Tempdir::with_prefix(...).unwrap_or("/tmp/default-cache-failed-tmpdir")...` (likely with some Rust type changes to thread the `PathBuf` type through. The downside is that path may not be unique and empty leading to other problems later on. 

I guess I'm now wishing for a `TryDefault` trait now!

Reply:
Yeah, my thinking is that since this is on boot, we likely want to fail here if we can't create the directories correctly as we probably will panic somewhere very shortly after this